File: src/ios/VpnDetector.m
#import "VpnDetector.h"
#import <ifaddrs.h>
#import <net/if.h>

@implementation VpnDetector

- (void)isVpnActive:(CDVInvokedUrlCommand*)command {
    BOOL vpnActive = [self checkVPN];
    
    CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK 
                                                   messageAsBool:vpnActive];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (BOOL)checkVPN {
    struct ifaddrs *interfaces = NULL;
    struct ifaddrs *temp_addr = NULL;
    BOOL vpnActive = NO;

    if (getifaddrs(&interfaces) == 0) {
        temp_addr = interfaces;
        while (temp_addr != NULL) {
            NSString *name = [NSString stringWithUTF8String:temp_addr->ifa_name];
            if ([name hasPrefix:@"utun"] || [name hasPrefix:@"ppp"]) {
                vpnActive = YES;
                break;
            }
            temp_addr = temp_addr->ifa_next;
        }
    }

    if (interfaces != NULL) {
        freeifaddrs(interfaces);
    }

    return vpnActive;
}

@end

File: src/ios/VpnDetector.swift
import Foundation
import NetworkExtension
import Cordova

@objc(VpnDetector) class VpnDetector: CDVPlugin {

    @objc(isVpnActive:)
    func isVpnActive(command: CDVInvokedUrlCommand) {
        var isVPNConnected = false
        if let settings = CFNetworkCopySystemProxySettings()?.takeRetainedValue() as? [String: Any],
           let scopes = settings["__SCOPED__"] as? [String: Any] {
            for key in scopes.keys {
                if key.contains("tap") || key.contains("tun") || key.contains("ppp") {
                    isVPNConnected = true
                    break
                }
            }
        }

        let pluginResult = CDVPluginResult(status: .ok, messageAs: isVPNConnected)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
}
